### ISSUE TITLE
Heap out of memory error when using QueryBuilder insert

### ISSUE THREAD

**Issue by MaikelH** [typeorm/typeorm#1191]
### Problem
When doing the new tests for #1115 I added an extra test using the QueryBuilder insert function. This should provide better performance when inserting a large amount of documents. When I try insert a batch of 2000 documents the following error occurs.

```
<--- Last few GCs --->

   46125 ms: Mark-sweep 1126.8 (1438.7) -> 1129.9 (1438.7) MB, 340.5 / 0.0 ms [allocation failure] [scavenge might not succeed].
   46465 ms: Mark-sweep 1129.9 (1438.7) -> 1132.9 (1438.7) MB, 339.7 / 0.0 ms [allocation failure] [scavenge might not succeed].
   46805 ms: Mark-sweep 1132.9 (1438.7) -> 1142.0 (1422.7) MB, 340.2 / 0.0 ms [last resort gc].
   47147 ms: Mark-sweep 1142.0 (1422.7) -> 1151.1 (1422.7) MB, 341.3 / 0.0 ms [last resort gc].
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x7e2c08cf781 <JS Object>
    1: StringReplaceGlobalRegExpWithFunction [native regexp.js:~460] [pc=0x20abfc31c483] (this=0x39c16f19f3b9 <JS Global Object>,W=0x318428433ba9 <Very long string[285428]>,C=0x3184284bf1a9 <JS RegExp>,aw=0x3184284bf1e1 <JS Function (SharedFunctionInfo 0x324dbc7d0ef1)>)
    2: [Symbol.replace] [native regexp.js:568] [pc=0x20abfc38d94d] (this=0x3184284bf1a9 <JS RegExp>,D=0x318428433ba9 <Very long...

 1: node::Abort() [/usr/bin/node]
 2: 0x109f13c [/usr/bin/node]
 3: v8::Utils::ReportApiFailure(char const*, char const*) [/usr/bin/node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/bin/node]
 5: v8::internal::Factory::NewFixedArray(int, v8::internal::PretenureFlag) [/usr/bin/node]
 6: 0xeea147 [/usr/bin/node]
 7: v8::internal::Runtime_RegExpExecMultiple(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/bin/node]
 8: 0x20abfb5092a7

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```
A workaround is to increase the heap of the NodeJS process. This is of course not a long term solution. It seems a lot of extra memory is allocated when doing the insert.

### Environment
**Typescript**: 2.5.3
**Database**: 9.6
**Typeorm**: 0.2.0-alpha.1
**NodeJS**: 6.11.3

### Code

```typescript
    public static start(docs: Document[]) : Promise<any> {
        let db = createConnection({
            type: "postgres",
            host: "localhost",
            port: 5432,
            username: "postgres",
            password: "test",
            database: "perf_test",
            entities: [
                Document
            ]
        }).catch(console.error);

        // Using insert on QueryBuilder (see comment by @pleerock
        // https://github.com/typeorm/typeorm/issues/1115#issuecomment-343125550
        return db.then((con: Connection) => {
            let start = new Date().getTime();

            return con.createQueryBuilder()
                .insert()
                .into(Document)
                .values(docs)
                // .useTransaction(true) // completely optional since now you must have only one insertion query with bulk insertion
                .execute()
                .then(doc => {
                    let end = new Date().getTime();

                    console.log("[TypeormInsert] Call to persist took " + (end - start) + " milliseconds.");
                })
                .then(() => {
                    return con.close();
                });
        });
    }
```

**Comment by pleerock** [comment]
I can't reproduce memory error on my 8gb ram mac. Here are my execution results:

```
### Run 1
[PG-Batch] Call to persist took 1162 milliseconds.
[TypeormInsert] Call to persist took 46354 milliseconds.
[TypeormSave] Call to persist took 52957 milliseconds.
[PG-InsertHelper] Call to persist took 861 milliseconds.
Finished test.
```

As you can see it still takes too much time. I found a bottleneck and fixed it in `0.2.0-alpha.2` and now it takes only:

```
### Run 1
[PG-Batch] Call to persist took 1067 milliseconds.
[TypeormInsert] Call to persist took 3156 milliseconds.
[TypeormSave] Call to persist took 5351 milliseconds.
[PG-InsertHelper] Call to persist took 227 milliseconds.
Finished test.
```

**Comment by pleerock** [comment]
Let me close this issue and continue discussion on our previously opened issue. I want to keep all our discussion history in a single ticket.
